File: apps/backend/apps/banking/serializers/bank_account.py
"""
Bank Account Serializers for LedgerSG Banking Module.

Validates all bank account data before persistence.
SEC-001 Remediation: Replaces unvalidated request.data.get() stubs.
"""

from decimal import Decimal
from rest_framework import serializers
from django.utils.translation import gettext_lazy as _

from apps.core.models import BankAccount, Account


class BankAccountSerializer(serializers.ModelSerializer):
    """Read serializer for BankAccount."""

    class Meta:
        model = BankAccount
        fields = [
            "id",
            "org",
            "account_name",
            "bank_name",
            "account_number",
            "bank_code",
            "branch_code",
            "currency",
            "gl_account",
            "paynow_type",
            "paynow_id",
            "is_default",
            "is_active",
            "opening_balance",
            "opening_balance_date",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "org", "created_at", "updated_at"]


class BankAccountCreateSerializer(serializers.ModelSerializer):
    """
    Write serializer for creating BankAccount.

    Validates:
    - GL account belongs to org and is bank account type
    - PayNow validation: if paynow_type set, paynow_id required
    - No duplicate account_number per org
    """

    class Meta:
        model = BankAccount
        fields = [
            "account_name",
            "bank_name",
            "account_number",
            "bank_code",
            "branch_code",
            "currency",
            "gl_account",
            "paynow_type",
            "paynow_id",
            "is_default",
            "opening_balance",
            "opening_balance_date",
        ]

    def validate_account_name(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError(_("Account name cannot be blank."))
        return value.strip()

    def validate_bank_name(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError(_("Bank name cannot be blank."))
        return value.strip()

    def validate_account_number(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError(_("Account number cannot be blank."))
        return value.strip()

    def validate_currency(self, value):
        if not value or len(value) != 3:
            raise serializers.ValidationError(_("Currency must be a 3-letter code."))
        return value.upper()

    def validate_opening_balance(self, value):
        if value is None:
            return Decimal("0")
        if value < Decimal("0"):
            raise serializers.ValidationError(_("Opening balance cannot be negative."))
        return value

    def validate_gl_account(self, value):
        org_id = self.context.get("org_id")
        if not org_id:
            raise serializers.ValidationError(_("Organisation context required."))

        try:
            account = Account.objects.get(id=value.id, org_id=org_id)
        except Account.DoesNotExist:
            raise serializers.ValidationError(_("GL account not found in this organisation."))

        account_type_name = account.account_type.upper() if account.account_type else ""
        if account_type_name != "ASSET":
            raise serializers.ValidationError(
                _("GL account must be an Asset account for bank accounts.")
            )

        sub_type_name = ""
        if account.account_sub_type:
            sub_type_name = account.account_sub_type.name.upper()

        valid_sub_types = ["BANK", "CASH AND BANK", "CASH & BANK", "CASH"]
        if sub_type_name and sub_type_name not in valid_sub_types:
            raise serializers.ValidationError(
                _("GL account must be a Bank or Cash type asset account.")
            )

        return value

    def validate(self, data):
        paynow_type = data.get("paynow_type")
        paynow_id = data.get("paynow_id")

        if paynow_type and not paynow_id:
            raise serializers.ValidationError(
                {"paynow_id": _("PayNow ID is required when PayNow type is specified.")}
            )

        if paynow_id and not paynow_type:
            raise serializers.ValidationError(
                {"paynow_type": _("PayNow type is required when PayNow ID is provided.")}
            )

        if paynow_id:
            paynow_type_value = paynow_type if isinstance(paynow_type, str) else paynow_type[0]
            if paynow_type_value == "UEN" and len(paynow_id) > 10:
                raise serializers.ValidationError(
                    {"paynow_id": _("UEN PayNow ID must be 10 characters or less.")}
                )
            elif paynow_type_value == "MOBILE" and not paynow_id.startswith("+"):
                raise serializers.ValidationError(
                    {"paynow_id": _("Mobile PayNow ID must start with '+' (e.g., +65XXXXXXXX).")}
                )
            elif paynow_type_value == "NRIC" and len(paynow_id) != 9:
                raise serializers.ValidationError(
                    {"paynow_id": _("NRIC PayNow ID must be exactly 9 characters.")}
                )

        return data


class BankAccountUpdateSerializer(serializers.ModelSerializer):
    """Write serializer for updating BankAccount."""

    class Meta:
        model = BankAccount
        fields = [
            "account_name",
            "bank_name",
            "bank_code",
            "branch_code",
            "paynow_type",
            "paynow_id",
            "is_default",
            "is_active",
            "opening_balance",
            "opening_balance_date",
        ]

    def validate_account_name(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError(_("Account name cannot be blank."))
        return value.strip()

    def validate_bank_name(self, value):
        if not value or not value.strip():
            raise serializers.ValidationError(_("Bank name cannot be blank."))
        return value.strip()

    def validate_opening_balance(self, value):
        if value is not None and value < Decimal("0"):
            raise serializers.ValidationError(_("Opening balance cannot be negative."))
        return value

    def validate(self, data):
        paynow_type = data.get("paynow_type")
        paynow_id = data.get("paynow_id")

        if paynow_type and not paynow_id:
            raise serializers.ValidationError(
                {"paynow_id": _("PayNow ID is required when PayNow type is specified.")}
            )

        if paynow_id and not paynow_type:
            raise serializers.ValidationError(
                {"paynow_type": _("PayNow type is required when PayNow ID is provided.")}
            )

        return data

File: apps/web/src/shared/schemas/bank-account.ts
import { z } from "zod";

/**
 * LEDGERSG BANK ACCOUNT SCHEMA
 *
 * IRAS Compliance: PayNow validation for Singapore businesses
 * Currency: 3-letter ISO code
 * Financial Precision: 4 decimal places (internal)
 *
 * Backend Reference: apps/backend/apps/banking/serializers/bank_account.py
 */

// PayNow Types (Singapore-specific)
export const PAYNOW_TYPES = ["UEN", "MOBILE", "NRIC"] as const;
export type PayNowType = (typeof PAYNOW_TYPES)[number];

/**
 * Bank Account Schema (Read - from API)
 *
 * Matches BankAccountSerializer from backend
 */
export const bankAccountSchema = z
  .object({
    id: z.string().uuid(),
    org: z.string().uuid(),
    account_name: z.string().min(1).max(200),
    bank_name: z.string().min(1).max(200),
    account_number: z.string().min(1).max(50),
    bank_code: z.string().max(10).optional().nullable(),
    branch_code: z.string().max(10).optional().nullable(),
    currency: z.string().length(3).toUpperCase(),
    gl_account: z.string().uuid(),
    paynow_type: z.enum(PAYNOW_TYPES).optional().nullable(),
    paynow_id: z.string().max(20).optional().nullable(),
    is_default: z.boolean(),
    is_active: z.boolean(),
    opening_balance: z.string().regex(/^\d+\.\d{4}$/), // 4dp internal precision
    opening_balance_date: z.string().optional().nullable(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .refine(
    (data) => {
      // PayNow type and ID must both be set or both be null
      if (data.paynow_type && !data.paynow_id) return false;
      if (data.paynow_id && !data.paynow_type) return false;
      return true;
    },
    {
      message: "PayNow type and ID must both be set or both be null",
    }
  )
  .refine(
    (data) => {
      if (!data.paynow_type || !data.paynow_id) return true;

      // Validate PayNow ID format based on type
      switch (data.paynow_type) {
        case "UEN":
          return data.paynow_id.length <= 10;
        case "MOBILE":
          return data.paynow_id.startsWith("+");
        case "NRIC":
          return data.paynow_id.length === 9;
        default:
          return true;
      }
    },
    {
      message: "Invalid PayNow ID format for type",
    }
  );

/**
 * Bank Account Input Schema (Write - to API)
 *
 * Matches BankAccountCreateSerializer from backend
 */
export const bankAccountInputSchema = z
  .object({
    account_name: z.string().min(1, "Account name required").max(200),
    bank_name: z.string().min(1, "Bank name required").max(200),
    account_number: z.string().min(1, "Account number required").max(50),
    bank_code: z.string().max(10).optional(),
    branch_code: z.string().max(10).optional(),
    currency: z.string().length(3).toUpperCase().default("SGD"),
    gl_account: z.string().uuid({ message: "GL account must be a valid UUID" }),
    paynow_type: z.enum(PAYNOW_TYPES).optional().nullable(),
    paynow_id: z.string().max(20).optional().nullable(),
    is_default: z.boolean().optional().default(false),
    opening_balance: z
      .string()
      .regex(/^\d*\.?\d{0,4}$/, "Invalid amount format")
      .default("0.0000"),
    opening_balance_date: z.string().optional().nullable(),
  })
  .refine(
    (data) => {
      // PayNow type and ID must both be set or both be null
      if (data.paynow_type && !data.paynow_id) return false;
      if (data.paynow_id && !data.paynow_type) return false;
      return true;
    },
    {
      message: "PayNow type and ID must both be set or both be null",
      path: ["paynow_id"],
    }
  )
  .refine(
    (data) => {
      if (!data.paynow_type || !data.paynow_id) return true;

      // Validate PayNow ID format based on type
      switch (data.paynow_type) {
        case "UEN":
          return data.paynow_id.length <= 10;
        case "MOBILE":
          return data.paynow_id.startsWith("+");
        case "NRIC":
          return data.paynow_id.length === 9;
        default:
          return true;
      }
    },
    {
      message: "Invalid PayNow ID format for type",
      path: ["paynow_id"],
    }
  );

/**
 * Bank Account Update Schema (Write - to API)
 *
 * Matches BankAccountUpdateSerializer from backend
 */
export const bankAccountUpdateSchema = z.object({
  account_name: z.string().min(1).max(200).optional(),
  bank_name: z.string().min(1).max(200).optional(),
  bank_code: z.string().max(10).optional().nullable(),
  branch_code: z.string().max(10).optional().nullable(),
  paynow_type: z.enum(PAYNOW_TYPES).optional().nullable(),
  paynow_id: z.string().max(20).optional().nullable(),
  is_default: z.boolean().optional(),
  is_active: z.boolean().optional(),
  opening_balance: z.string().regex(/^\d*\.?\d{0,4}$/).optional(),
  opening_balance_date: z.string().optional().nullable(),
});

// Export types
export type BankAccount = z.infer<typeof bankAccountSchema>;
export type BankAccountInput = z.infer<typeof bankAccountInputSchema>;
export type BankAccountUpdate = z.infer<typeof bankAccountUpdateSchema>;

/**
 * Factory: Create empty bank account input with defaults
 */
export function createEmptyBankAccountInput(): BankAccountInput {
  return {
    account_name: "",
    bank_name: "",
    account_number: "",
    currency: "SGD",
    gl_account: "",
    is_default: false,
    opening_balance: "0.0000",
  };
}

/**
 * Helper: Format bank account for display
 */
export function formatBankAccountName(account: BankAccount): string {
  return `${account.bank_name} - ${account.account_number}`;
}

/**
 * Helper: Format opening balance for display (2dp)
 */
export function formatOpeningBalance(balance: string): string {
  const num = parseFloat(balance);
  return num.toLocaleString("en-SG", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}
